Add profiling around one timestep

brryan opened the pull request:
## Background

We want to be able to cleanly see one Jaybenne timestep in profiling tools

## Description of Changes

- [x] Add `Kokkos::Profiling::pushRegion("Jaybenne::Timestep"); at start of first task in Jaybenne timestep
- [x] Add `Kokkos::Profiling::popRegion();` at end of last task in Jaybenne timestep

Note that, because both these tasks are inside loops over partitions, this may not work (I'm not sure) if we have more than one partition. But right now we require one partition anyway so that's not an immediate issue.

## Checklist

- [ ] New features are documented
- [x] Tests added for bug fixes and new features
- [x] (@lanl.gov employees) Update copyright on changed files

Commit:
Add push/pop profiling region around entire timestep

brryan commented on src/jaybenne/jaybenne.cpp:
```diff
@@ -573,6 +575,8 @@ TaskStatus UpdateFluid(MeshData<Real> *md) {
         ee += delta;
       });
 
+  Kokkos::Profiling::popRegion(); // End of Jaybenne timestep
```
@fglines-nv Am I correct that we don't need a preceding `Kokkos::fence()` here because when using a profiling library like nvtx connector, `popRegion` will include its own `fence()` call?

fglines-nv replied:
We don't need a `Kokkos::fence()`, that's correct. There won't be any `fence()` or other host-device synchronization points with the `nvtx` connector. `nsys` will capture the timeline of execution on both the CPU and GPU, which will be disconnected.

pdmullen reviewed (approved):
This LGTM! Is this actually targeting `develop` or just a working branch for performance diagnostics? 

If the former, maybe we could add a comment around the `Kokkos::Profile::pushRegion` calls so that users/devs know where the matching starting/closing calls are... or maybe, we could add two new tasks to the driver (optionally added by a `if (perf_test)`) that call the beginning and ending `Kokkos::Profile::pushRegion`s.  

Please consider neither of these blocking.

pdmullen commented:
Another thought: `UpdateFluid` may not be called by some `jaybenne` configurations (e.g., `stepdiff`).

brryan committed:
Not right fixes for this PR

brryan committed:
Oops bad submodule

brryan commented:
> This LGTM! Is this actually targeting `develop` or just a working branch for performance diagnostics?
> 
> If the former, maybe we could add a comment around the `Kokkos::Profile::pushRegion` calls so that users/devs know where the matching starting/closing calls are... or maybe, we could add two new tasks to the driver (optionally added by a `if (perf_test)`) that call the beginning and ending `Kokkos::Profile::pushRegion`s.
> 
> Please consider neither of these blocking.

Yeah I think we want this to go into develop. 

I followed your advice and moved the `push`/`pop` calls out of existing functions and into their own tasks. This also fixes the multiple partitions issue. 

> Another thought: `UpdateFluid` may not be called by some `jaybenne` configurations (e.g., `stepdiff`).

Yes this was a bug, thanks for catching this. Fixed now.

fglines-nv commented:
> Yeah I think we want this to go into develop.

These are no-OPs when profiling is turned off, they should be fine in develop

> I followed your advice and moved the push/pop calls out of existing functions and into their own tasks. This also fixes the multiple partitions issue.

Adding these as their own tasks seems very convoluted, what was the issue with having them tightly coupled to the functions they're profiling? 

What was the issue with multiple partitions? For CUDA, these would translate to multiple streams which should work fine with the same NVTX region names across different streams.

One more profiling region request -- can we add a profiling region around each iteration of `TransportPhotons` and `TransportPhotons_DDMC`? I haven't tracked down exactly where that iteration happens

pdmullen commented:
> > Yeah I think we want this to go into develop.
> 
> These are no-OPs when profiling is turned off, they should be fine in develop

Totally agree 😸 I just didn't want my comments to be more annoying than anything if this was just a working branch. 

>
> > I followed your advice and moved the push/pop calls out of existing functions and into their own tasks. This also fixes the multiple partitions issue.
> 
> Adding these as their own tasks seems very convoluted, what was the issue with having them tightly coupled to the functions they're profiling?
> 
> What was the issue with multiple partitions? For CUDA, these would translate to multiple streams which should work fine with the same NVTX region names across different streams.

The concern here was that their earlier placement was not indicative of the profiler testing the whole `Jaybenne` timestep.  We were putting `push` towards the beginning of a first expected task, and the `pop` during the last expected task (which may very depending on configuration).  I proposed to either comment this, or rather, to surround the entire driver via the `push`/`pop` duo.  I think @brryan opted for the latter (which is nice in that it also fixes issues with partitions, since the driver can potentially be executed over `num_partitions == nblocks_per_rank` when `pack_size==1`)


> 
> One more profiling region request -- can we add a profiling region around each iteration of `TransportPhotons` and `TransportPhotons_DDMC`? I haven't tracked down exactly where that iteration happens

Thanks for working on this with us!

fglines-nv commented:
> The concern here was that their earlier placement was not indicative of the profiler testing the whole `Jaybenne` timestep. We were putting `push` towards the beginning of a first expected task, and the `pop` during the last expected task (which may very depending on configuration). I proposed to either comment this, or rather, to surround the entire driver via the `push`/`pop` duo. I think @brryan opted for the latter (which is nice in that it also fixes issues with partitions, since the driver can potentially be executed over `num_partitions == nblocks_per_rank` when `pack_size==1`)

I think I understand -- "one step" really isn't a function in Jaybenne since the iteration over timesteps/the task list is in Parthenon.

This makes me think that this profiling region belongs in Parthenon instead. I'll look at that

pdmullen commented:
> > The concern here was that their earlier placement was not indicative of the profiler testing the whole `Jaybenne` timestep. We were putting `push` towards the beginning of a first expected task, and the `pop` during the last expected task (which may very depending on configuration). I proposed to either comment this, or rather, to surround the entire driver via the `push`/`pop` duo. I think @brryan opted for the latter (which is nice in that it also fixes issues with partitions, since the driver can potentially be executed over `num_partitions == nblocks_per_rank` when `pack_size==1`)
> 
> I think I understand -- "one step" really isn't a function in Jaybenne since the iteration over timesteps/the task list is in Parthenon.
> 
> This makes me think that this profiling region belongs in Parthenon instead. I'll look at that

Maybe we are saying the same thing... 😄  But just in case, to phrase a bit differently, we we want to profile the entire `Step` driver of `Jaybenne`.  The driver invokes a `TaskCollection` that can operate over `num_partitions`.  The design @brryan opted for allows us to fully envelop the `Driver` via the `push`/`pop` duo.  The `Step` driver does indeed call `Parthenon` functions such as boundary exchange.

brryan committed:
Profiling around transport iteration, fix inf tests

brryan commented:
> > > The concern here was that their earlier placement was not indicative of the profiler testing the whole `Jaybenne` timestep. We were putting `push` towards the beginning of a first expected task, and the `pop` during the last expected task (which may very depending on configuration). I proposed to either comment this, or rather, to surround the entire driver via the `push`/`pop` duo. I think @brryan opted for the latter (which is nice in that it also fixes issues with partitions, since the driver can potentially be executed over `num_partitions == nblocks_per_rank` when `pack_size==1`)
> > 
> > 
> > I think I understand -- "one step" really isn't a function in Jaybenne since the iteration over timesteps/the task list is in Parthenon.
> > This makes me think that this profiling region belongs in Parthenon instead. I'll look at that
> 
> Maybe we are saying the same thing... 😄 But just in case, to phrase a bit differently, we we want to profile the entire `Step` driver of `Jaybenne`. The driver invokes a `TaskCollection` that can operate over `num_partitions`. The design @brryan opted for allows us to fully envelop the `Driver` via the `push`/`pop` duo. The `Step` driver does indeed call `Parthenon` functions such as boundary exchange.

I think we can still effectively capture the "one step" profiling on the Jaybenne side with the push and pop at beginning and end of the task collection jaybenne creates. Rather than doing this in Parthenon, I think it makes sense to include that profiling logic in Jaybenne here because if we run jaybenne from artemis, we will still get a faithful report of only the time spent in the jaybenne part of the timestep cycle. It might be nice to also have Parthenon return per-cycle and per-integrator-stage profiling information which like you say is probably best implemented in Parthenon. 

I also added a profiling region with more separate tasks inside each iteration of the transport loop. Personally I'm OK with this extra verbosity in the construction of the task list itself, since as Patrick says it does help with seeing immediately where things start and stop.

fglines-nv reviewed (approved):
![image](https://github.com/user-attachments/assets/0eadc85d-e8cc-47ba-b7f6-837a764319be)

This is working well, looks good!

brryan commented:
> ![image](https://private-user-images.githubusercontent.com/165137654/389709028-0eadc85d-e8cc-47ba-b7f6-837a764319be.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NzA4MTEsIm5iZiI6MTczMjU3MDUxMSwicGF0aCI6Ii8xNjUxMzc2NTQvMzg5NzA5MDI4LTBlYWRjODVkLWU4Y2MtNDdiYS1iN2Y2LTgzN2E3NjQzMTliZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNVQyMTM1MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNDc2NjE5ZTZhZGY2ZWM2YzUxMDRmMjA2Yjg3OTJjODFjNTIzNmY0ODQwNGQyNjQ5YmY5YjkyZTE4OTk5Y2FjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.uPFpVWYIzNsMztIWJCpbC6RwYZxq5DlRaIY7HyucwGI)
> 
> This is working well, looks good!

Great, glad it works!
